apps/prairielearn/src/pages/instructorAssessmentAccessOverrides/instructorAssessmentAccessOverrides.js

```diff
@@ -0,0 +1,149 @@
+// @ts-check
+import * as express from 'express';
+const asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import * as error from '@prairielearn/error';
+import { getEnrollmentForUserInCourseInstance } from '../../models/enrollment';
+import { selectUserByUid } from '../../models/user';
+
+const router = express.Router();
+const sql = sqldb.loadSqlEquiv(__filename);
+
+async function getUserIdAndCheckEnrollment({ uid, course_instance_id }) {
+  const user = await selectUserByUid(uid);
+  if (!user) {
+    throw error.make(400, `User ${uid} does not exist.`);
+  }
+
+  const enrollment = await getEnrollmentForUserInCourseInstance({
+    user_id: user.user_id,
+    course_instance_id: course_instance_id,
+  });
+  if (!enrollment) {
+    throw error.make(400, `User ${uid} is not enrolled in this course instance.`);
+  }
+
+  return user.user_id;
+}
```

Comment:
I think these comments got lost after a recent change, so I'll repeat them here:

This is a regression from the fix for the problem discussed in https://github.com/PrairieLearn/PrairieLearn/pull/9103#discussion_r1496735902. With this, a user could enumerate which students are present in PrairieLearn.

Reply:
I think I understand, to clarify, I should check enrollment and then get user_id for params like the following
```javascript
if (req.body.student_uid) {
        const checkEnrollment = await userIsEnrolledInCourseInstance({
          uid: req.body.student_uid,
          course_instance_id: res.locals.course_instance.id,
        });
        if (!checkEnrollment) {
          throw error.make(400, `User ${req.body.student_uid} is not enrolled in this course instance.`);
        }
        const user = await selectUserByUid(req.body.student_uid);
        user_id = user ? user.user_id : null;
      }
```

Reply:
Ok, I think I understand. So in order to not allow users to enumerate, we should check enrollment and then get user_id correct? so the following: 
```
if (req.body.student_uid) {
        const checkEnrollment = await userIsEnrolledInCourseInstance({
          uid: req.body.student_uid,
          course_instance_id: res.locals.course_instance.id,
        });
        if (!checkEnrollment) {
          throw error.make(400, `User ${req.body.student_uid} is not enrolled in this course instance.`);
        }
        const user = await selectUserByUid(req.body.student_uid);
        if (!user) {
          throw error.make(400, `User ${req.body.student_uid} does not exist.`);
        }
        user_id = user.user_id;
      }
```

Also committed for ease of viewing: 77f87387e

Reply:
due to 'As written, this would allow someone to enumerate which students exist in PrairieLearn, since it differentiates between a user not existing and a user not being enrolled in the course instance.' [comment](https://github.com/PrairieLearn/PrairieLearn/pull/9103#discussion_r1496735902) made , I think I am still differentiating between a user not existing and a user not being enrolled. I am still doing this when I get user_id, and I am having trouble seeing how I would get user_id for the params without separating course enrollment and if a user exists. I will keep thinking about this but wanted to update where I was at.

Reply:
What you have now is looking reasonable. The key is that with your previous code, someone could differentiate between "this user exists but is not enrolled" and "this user does not exist".

However, with your changes, we now select the user twice, which isn't ideal. Consider instead introducing a function like `selectUserEnrolledInCourseInstance`, reading something like this:

```ts
async function selectUserEnrolledInCourseInstance({ uid, course_instance_id }) {
  const user = await selectUserByUid(uid);
  if (!user) return null;

  const enrollment = await getEnrollmentForUserInCourseInstance({
    user_id: user.user_id,
    course_instance_id: course_instance_id,
  });
  if (!enrollment) return null;

  return user;
}
```

Then, when you call it, you can perform just a single check:

```ts
const user = await selectUserEnrolledInCourseInstance(...);
if (!user) {
  throw error.make(400, `User ${req.body.student_uid} is not enrolled in this course instance.`);
}
user_id = user.user_id;
```

Reply:
Makes sense, done in a9330c3cb

Reply:
fixed the error message in 91c2bae40